MapSplitBrainStressTest.testSplitBrain

_maintenance_ (commit c88193f1f51ffa464bbf6867e61cbbb6a970a765)

Failed on nightly build : http://jenkins.hazelcast.com/job/Hazelcast-3.maintenance-nightly/59/testReport/junit/com.hazelcast.map.merge/MapSplitBrainStressTest/testSplitBrain/

**Stacktrace**
```
java.lang.AssertionError: Instances not in safe state! {[127.0.0.1]:5704=REPLICA_NOT_SYNC}
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertAllInSafeState(HazelcastTestSupport.java:817)
	at com.hazelcast.test.HazelcastTestSupport$4.run(HazelcastTestSupport.java:782)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1294)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1311)
	at com.hazelcast.test.HazelcastTestSupport.waitAllForSafeState(HazelcastTestSupport.java:780)
	at com.hazelcast.test.HazelcastTestSupport.waitAllForSafeState(HazelcastTestSupport.java:776)
	at com.hazelcast.test.HazelcastTestSupport.waitAllForSafeState(HazelcastTestSupport.java:802)
	at com.hazelcast.test.SplitBrainTestSupport.healSplitBrain(SplitBrainTestSupport.java:322)
	at com.hazelcast.test.SplitBrainTestSupport.doIteration(SplitBrainTestSupport.java:225)
	at com.hazelcast.test.SplitBrainTestSupport.testSplitBrain(SplitBrainTestSupport.java:214)
	at com.hazelcast.map.merge.MapSplitBrainStressTest.testSplitBrain(MapSplitBrainStressTest.java:89)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java
@@ -1095,6 +1095,17 @@ public class MembershipManager {
 
         @Override
         public void run() {
+            try {
+                innerRun();
+            } catch (Throwable e) {
+                logger.warning("Exception thrown while running DecideNewMembersViewTask", e);
+            } finally {
+                // Resume migrations, they are disabled when mastership claim is started
+                node.getPartitionService().resumeMigration();
+            }
+        }
+
+        private void innerRun() {
             MembersView newMembersView = decideNewMembersView(localMemberMap, membersToAsk);
             clusterServiceLock.lock();
             try {
@@ -1123,8 +1134,6 @@ public class MembershipManager {
                 sendMemberListToOthers();
                 logger.info("Mastership is claimed with: " + newMembersView);
             } finally {
-                // Resume migrations, they are disabled when mastership claim is started
-                node.getPartitionService().resumeMigration();
                 clusterServiceLock.unlock();
             }
         }

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
@@ -555,6 +555,8 @@ public class MigrationManager {
         migrationQueue.clear();
         activeMigrationInfo = null;
         completedMigrations.clear();
+        shutdownRequestedAddresses.clear();
+        migrationAllowed.set(true);
     }
 
     void start() {

